### DISCUSSION TITLE
Issue with Authentication in Multi-Tenancy Setup

### DISCUSSION THREAD

**Discussion by ousid** [filamentphp/filament#8714] [Help] [answered]
**GitHub Discussion: Issue with Authentication in Multi-Tenancy Setup**

**Introduction:**
I'm encountering an issue with the authentication process in my Laravel project, which involves several packages and a multi-tenancy setup. Here are the details of my environment:

- Filament `V3.x`
- Livewire `V3.x`
- Laravel `V10.x`
- [archtechx/tenancy](https://github.com/archtechx/tenancy) `V3`
- Using the `InitializeTenancyByDomainOrSubdomain::class` Middleware
- Employing Multi-Database Usage

**The Issue:**
When attempting to authenticate users, I've encountered a problem where users can log in successfully, but there's no authenticated session. Consequently, they get logged out and redirected back to the login page. After thorough investigation, I've determined that the culprit behind this behavior is the `Authenticate::class` middleware provided by Filament.

**Steps to Reproduce:**
To replicate this issue, I followed these steps:

1. Created a Laravel project.
2. Installed the Filament package.
3. Installed the Tenancy package.
4. Configured the Tenancy package to support multi-database setups.
5. Added a new panel to handle tenants.
6. Included the `InitializeTenancyByDomainOrSubdomain::class` middleware in the new `PanelProvider`.
7. Updated the Livewire update route.

**Code Samples:**
Here are relevant code snippets from my configuration:

Livewire Update Route (in `AppServiceProvider.php`):
```php
Livewire::setUpdateRoute(function ($handle) {
    return Route::post('/livewire/update', $handle)
        ->middleware(
            InitializeTenancyByDomainOrSubdomain::class
        );
});
```

Filament Panel Configuration (in `StorePanelProvider`):
```php
// ...
->middleware([
    EncryptCookies::class,
    AddQueuedCookiesToResponse::class,
    StartSession::class,
    AuthenticateSession::class,
    ShareErrorsFromSession::class,
    VerifyCsrfToken::class,
    SubstituteBindings::class,
    DisableBladeIconComponents::class,
    DispatchServingFilamentEvent::class,
    InitializeTenancyByDomainOrSubdomain::class,
    PreventAccessFromCentralDomains::class,
])
->authMiddleware([
    Authenticate::class,
]);
```

**Expected Behavior:**
Ideally, when accessing a new tenant like `store.example.com/store/login`, the following should happen:

1. The tenant initializes.
2. The database switches to the `tenant_store`.
3. The user gets authenticated based on the records associated with the mentioned tenant.
4. The login process works smoothly.
5. Access to the panel is granted.

**The Problem:**

The issue arises when trying to access the panel. The `Authenticate` middleware, while checking for an authenticated user, fails to find one. Consequently, it automatically "unauthenticates" the user and redirects them to the login page. 

__Removing the Livewire update route__ makes the process work, but it logs the user in from the central tenant instead of the current tenant.

I've spent two days troubleshooting this problem, searching online, and reviewing the source code, but haven't found a satisfactory solution. If anyone has encountered a similar issue or has any hints on how to resolve it, I would greatly appreciate the assistance.

Thank you!
cc: @danharrin, @stancl

**Comment by francoism90** [comment]
Shouldn't you register this in a global service provider instead?

We are using Spatie, and this works just fine.

**Reply by ousid** [reply]
Hey, @francoism90 thank you for your reply, but I'm afraid that I don't understand What do you mean by registering it in global service provider?

**Reply by francoism90** [reply]
@ousid https://laravel.com/docs/10.x/providers#writing-service-providers, and add this to `config/app.php`.

I think you should apply this when Livewire gets initialized, which should before Filament, like inside a `LivewireServiceProvider`, and use `Livewire::addPersistentMiddleware(..)`.

**Reply by ousid** [reply]
Thanks for the suggestion! I didn't know that you are about registering livewire middleware globally.

This will cause an expired page (419 error code). But I'm not sure why!

**Reply by ousid** [reply, marked as answer]
Well, the solution for this issue, is so funny to be honest :)

when I dive into the source code, of `Livewire::setUpdateRoute`, I found out, that it needs the `web` middleware, to authenticate the user correctly, so the solution would be, adding the `web` middleware along with the tenancy middleware 😆 

```php
        Livewire::setUpdateRoute(function ($handle) {
            return Route::post('/livewire/update', $handle)
                ->middleware(['web', InitializeTenancyByDomainOrSubdomain::class, ...]);
        });
```

Thank you for your help @francoism90.

**Reply by mbs047** [reply]
i know it's too late :), but i have this is helpfull

i worked around your code to make it works in both the central & tenants, and here is the updated code, feel free to change it / modify it


```php
// Extract the host (domain) from the current URL
$parsedUrl = parse_url(URL::current());
$host = $parsedUrl['host'];

// Log the current host for debugging purposes
\Log::info("Current Host: $host");

// Check if the current URL is not the same as the central domain
if ($host !== config('app.central_url')) {
    // Log that the condition is matched and middleware is being added
    \Log::info("Condition matched. Adding middleware.");

    // Configure the Livewire update route with middleware
    Livewire::setUpdateRoute(function ($handle) {
        return Route::post('/livewire/update', $handle)
            ->middleware([
                'web',
                InitializeTenancyByDomainOrSubdomain::class
            ]);
    });

    // Clear routes to ensure the new route is properly registered
    \Artisan::call('route:clear');
}
```

**Reply by nocodelab** [reply]
This one should be in the docs! Thanks mate! :)
This is the only way for me to make it work with central domain with Filament + tenant subdomains without it.

**Comment by jumawilliam** [comment]
Experiencing the same challenge but none of the above solutions working
